### ISSUE TITLE
EB 3.8.0 BUG?:  permissions on files in EASYBUILD_TMP_LOGDIR/reprod/easyblocks

### ISSUE THREAD

**Issue by JackPerdue** [easybuilders/easybuild-framework#2704]
I thought I'd try a random build of a new easyconfig using 3.8.0.  I went with Gerris/20131206-foss-2017b.  So I issued the command to do that and EB detected that it needed GTS as a dependency so it built it first.  But when it went to do the Gerris build proper it failed with:

```
== 2018-12-19 07:23:19,319 build_log.py:162 ERROR EasyBuild crashed with an error (at ?:124 in __init__): Failed to copy file /sw/eb/software/EasyBuild/3.8.0/lib/python2.6/site-packages/easybuild_easyblocks-3.8.0-py2.6.egg/easybuild/easyblocks/generic/configuremake.py to /tmp/j-perdue/logs/reprod/easyblocks/configuremake.py: [Errno 13] Permission denied: '/tmp/j-perdue/logs/reprod/easyblocks/configuremake.py' (at easybuild/tools/filetools.py:1673 in copy_file)
```

The problem here is that we use `EASYBUILD_READ_ONLY_INSTALLDIR=1` on all builds... including the one for EasyBuild/3.8.0.  So when GTS first copied the file there was no problem because configuremake.py wasn't in the reprod/easyblocks dir yet.  But when Gerris tried to copy the file it failed because the copy that GTS created in that dir was still marked as read-only.

Workaround was to add write-permission to the file and move forward.

Of course, after Gerris suceeded the file was left in a read-only state.

Other workaround for now is to not use READ_ONLY_INSTALLDIR when installing 3.8.0... but it would be better if fixed in 3.8.1.

**Comment by JackPerdue** [comment]
Actually.. I'll try sticking with the read-only install for now and just do a chmod u+w on all the .py files in the installed easyblocks dir.  Hopefully that will work (I think it should).

**Comment by boegel** [comment]
@JackPerdue This is strange, since the `reprod` dir is supposed to be unique for each installation, i.e. it's a subdirectory of `<installdir>/easybuild/`.

So the fact that installation directories of dependencies are read-only is not the problem here.

Can you share your EasyBuild configuration with us (output of `eb --show-config` should suffice) + the exact command you ran so we can try and reproduce this?

**Comment by ocaisa** [comment]
I don't fully follow the problem, from the complaint it says it failed to write to `/tmp/j-perdue/logs/reprod/easyblocks/configuremake.py`. I think the problem is that you seem to be using a static temp directory to hold your build logs, I guess I just didn't expect that.

**Comment by boegel** [comment]
cc @ocaisa

**Comment by ocaisa** [comment]
When we make the initial copy in the tmpdir, we could change the permissions at that point

**Comment by JackPerdue** [comment]
That would work... the problem is in the tmp dir not the final install dir.

[j-perdue@login2 (08:03) ~]$ eb --show-config
#
# Current EasyBuild configuration
# (C: command line argument, D: default value, E: environment variable, F: configuration file)
#
buildpath            (E) = /tmp/j-perdue/easybuild/
containerpath        (E) = /sw/eb/containers
debug                (E) = True
deprecated           (E) = 2.0
github-user          (E) = JackPerdue
include-easyblocks   (E) = /sw/eb/hprc/ec-dev/easyblocks/*.py
installpath          (E) = /sw/eb
minimal-toolchains   (E) = True
packagepath          (E) = /sw/eb/packages
prefix               (E) = /sw/eb
read-only-installdir (E) = True
repositorypath       (E) = /sw/eb/ebfiles_repo/ada
robot                (E) = /sw/eb/ebfiles_repo_cleaned/ada, /sw/eb/software/EasyBuild/3.8.0/lib/python2.6/site-packages/easybuild_easyconfigs-3.8.0-py2.6.egg/easybuild/easyconfigs
robot-paths          (E) = /sw/eb/ebfiles_repo_cleaned/ada, /sw/eb/software/EasyBuild/3.8.0/lib/python2.6/site-packages/easybuild_easyconfigs-3.8.0-py2.6.egg/easybuild/easyconfigs
sourcepath           (E) = /sw/eb/sources
tmp-logdir           (E) = /tmp/j-perdue/logs/

Command was:
eb /sw/eb/software/EasyBuild/3.8.0/lib/python2.6/site-packages/easybuild_easyconfigs-3.8.0-py2.6.egg/easybuild/easyconfigs/g/Gerris/Gerris-20131206-foss-2017b.eb

Doing chmod u+w on easyblocks in the install seems to have worked (I repeated the build from scratch for GTS/Gerris and had no problem).

**Comment by ocaisa** [comment]
Ok, yeah it's
```
tmp-logdir (E) = /tmp/j-perdue/logs/
```
that is the problem. With that setting different runs are not independent. I should've been more careful to cleanup the tmpdir *or* make sure that everything in there was writable. The way things are though, this is going to lead to potential leakage into the `reprod` directories of your installations (from failed builds or doing 2 builds at once...even if I fix things)

@boegel Do you think it's better to delete the directory if it already exists in the tmp-dir?

**Comment by ocaisa** [comment]
Or I change the `reprod` dir name to include  a timestamp when it's in the tmp space

**Comment by JackPerdue** [comment]
I certainly wouldn't want to preclude simultaneous builds so would go with the PID or timestamp.  I was wondering about different builds overwriting the easyblock and getting the final reprod wrong.   e.g. I could do one build using stock configuremake.py and then startup a build using my modified configuremake.py.... which build would get which copy in the current setup?

**Comment by JackPerdue** [comment]
I guess I forgot the last line from --show-config:

tmpdir               (E) = /tmp/j-perdue/tmp/

You could use the weird/confusing hash they use in that directory instead of the PID or timestamp. (???)

**Comment by ocaisa** [comment]
We can decide the convention in #2705

**Comment by boegel** [comment]
@JackPerdue I feel obliged to mention that this is sort of self-inflicted (which doesn't mean we won't fix this, it's still a bug, but it sort of explains why we overlooked this).

If `tmp-logdir` is not specified, you will basically get the temporary log file in a unique subdirectory of `/tmp/j-perdue/tmp/`, something like `/tmp/j-perdue/tmp/eb-iILVPp/`.

In that case, there's no problem at all, since the `reprod` subdirectory would be parked in there, so it'll be unique for each `eb` session/installation.

**Comment by boegel** [comment]
fixed in #2705, thanks for reporting @JackPerdue!
